sw/device/lib/dif/dif_kmac_unittest.cc

```diff
@@ -86,5 +92,77 @@ TEST(FunctionNameTest, BadArg) {
   EXPECT_EQ(dif_kmac_function_name_init("", -1, &fn), kDifBadArg);
 }
 
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Test;
+
+// Base class for the rest fixtures in this file.
+class KmacTest : public testing::Test, public mock_mmio::MmioTest {
+ public:
+  dif_kmac_t kmac_;
+  dif_kmac_operation_state_t kmac_operation_state_;
+
+  KmacTest() { EXPECT_EQ(dif_kmac_init(dev().region(), &kmac_), kDifOk); }
+
+  void setExpectedMessageByte(const uint8_t *message, const size_t size = 8) {
+    for (int i = 0; i < size; ++i) {
+      EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, message[i]);
+    }
+  }
+
+  void setExpectedMessageInt32(const uint8_t *message, const size_t size = 8) {
+    size_t remaining = size;
+    size_t unAlignement = ((uintptr_t)message) % sizeof(uint32_t);
+    if (unAlignement) {
+      setExpectedMessageByte(message, unAlignement);
+      message += unAlignement;
+      remaining -= unAlignement;
+    }
+
+    while (remaining > 3) {
+      uint32_t word = 0;
+      memcpy(&word, message, sizeof(uint32_t));
+      EXPECT_WRITE32(KMAC_MSG_FIFO_REG_OFFSET, word);
+      remaining -= sizeof(uint32_t);
+      message += sizeof(uint32_t);
+    }
+    if (remaining > 0) {
+      setExpectedMessageByte(message, remaining);
+    }
+  }
+};
+
+TEST_F(KmacTest, AbsorbAlignedMessage) {
+  COMPILER_RT_ALIGNAS(sizeof(uint32_t))
+  static const uint8_t kMessage[] = {0xa7, 0x48, 0x47, 0x93, 0x0a, 0x03,
+                                     0xab, 0xee, 0xa4, 0x73, 0xe1, 0xf3,
+                                     0xdc, 0x30, 0xb8, 0x88, 0x15};
+
+  const size_t kLen = sizeof(kMessage);
+
+  EXPECT_READ32(KMAC_STATUS_REG_OFFSET, 3);
+  setExpectedMessageInt32(kMessage, kLen);
+
+  EXPECT_EQ(
+      dif_kmac_absorb(&kmac_, &kmac_operation_state_, kMessage, kLen, NULL),
+      kDifOk);
+}
+
+TEST_F(KmacTest, AbsorbUnalignedMessage) {
+  COMPILER_RT_ALIGNAS(sizeof(uint16_t))
```

Comment:
You might be better off declaring the array module scope and using `__attribute__((aligned(2)))`.